### PR TITLE
[deps] Bump native assets dependencies

### PR DESCRIPTION
**Pull request by dcharkes**
Roll of a bunch of breaking changes from the native_assets_builder and native_assets_cli upstream. Most notably:

* https://github.com/dart-lang/native/pull/946
* https://github.com/dart-lang/native/pull/1018
* https://github.com/dart-lang/native/pull/1019

This PR also updates the template in `flutter create --template=package_ffi` to use the rewritten API.

This PR does not change any functionality in Flutter.

For reference, the same roll in the Dart SDK:

* https://dart-review.googlesource.com/c/sdk/+/357605
* https://dart-review.googlesource.com/c/sdk/+/357623

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart [hunk 1]
```diff
@@ -189,36 +186,37 @@ AndroidArch _getAndroidArch(Target target) {
...
       asset: _targetLocationAndroid(asset),
   };
 }
 
 /// Converts the `path` of [asset] as output from a `build.dart` invocation to
 /// the path used inside the Flutter app bundle.
-KernelAsset _targetLocationAndroid(Asset asset) {
-  final AssetPath path = asset.path;
+KernelAsset _targetLocationAndroid(AssetImpl asset) {
+  final LinkModeImpl linkMode = (asset as NativeCodeAssetImpl).linkMode;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart:200]: Out of curiosity, why is the tool now using *Impl types? I suppose since we're already importing an *internal library it doesn't matter also using internal implementation interfaces.

File: packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart [hunk 2]
```diff
@@ -189,36 +186,37 @@ AndroidArch _getAndroidArch(Target target) {
...
     default:
       throw Exception(
-        'Unsupported asset path type ${path.runtimeType} in asset $asset',
+        'Unsupported asset link mode $linkMode in asset $asset',
       );
   }
   return KernelAsset(
     id: asset.id,
-    target: asset.target,
+    target: Target.fromArchitectureAndOS(asset.architecture!, asset.os),
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart:219]: why was this change necessary?

**Review by christopherfujino — APPROVED** [review]
A few questions, mostly out of curiosity, but this LGTM once CI passes.

File: packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart (hunk 1 above)
Comment (dcharkes) [reply packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart:200]: The interfaces don't expose the APIs needed. (JSON serialization for example.) That's why we use the implementations.

The API is meant for `hook/built.dart` writers. The `*Impl` classes are meant for `hook/build.dart` invokers (`dartdev` and `flutter_tools`).

File: packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart (hunk 2 above)
Comment (dcharkes) [reply packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart:219]: `Target` is removed from the `hook/build.dart` API using `Asset`s, but still used in `KernelAsset`. `Target` is a VM concept, `hook/build.dart` writers will only talk about `OS` and `Architecture`. `dartdev` and `flutter_tools` have to do the conversion.

**Comment by dcharkes** [comment]
> A few questions, mostly out of curiosity, but this LGTM once CI passes.

Thanks @christopherfujino!

File: packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart:219]: Thanks
